### PR TITLE
Cost546 split payload

### PR DESCRIPTION
**Pull request by abaiken**
### 

### Testing 
1. Scenario with files that are not larger than the max size (I let the operator generate these files fore me): 
```
ashleybrookeaiken@aaiken-mac korekuta-operator-go % ls -lh /tmp/cost-mgmt-operator-reports/data 
total 1240
-rw-r--r--  1 ashleybrookeaiken  wheel    70K Oct 20 14:04 cm-openshift-namespace-labels-lookback-202010.csv
-rw-r--r--  1 ashleybrookeaiken  wheel    19K Oct 20 14:04 cm-openshift-node-labels-lookback-202010.csv
-rw-r--r--  1 ashleybrookeaiken  wheel   2.2K Oct 20 14:04 cm-openshift-persistentvolumeclaim-lookback-202010.csv
-rw-r--r--  1 ashleybrookeaiken  wheel   420K Oct 20 14:04 cm-openshift-usage-lookback-202010.csv
ashleybrookeaiken@aaiken-mac korekuta-operator-go %
```
Logs: 
```
2020-10-23T15:36:30.387-0400    INFO    controllers.CostManagement      There have been no prior successful uploads to cloud.redhat.com.        {"costmanagement": "checkCycle"}
2020-10-23T15:36:30.387-0400    INFO    controllers.CostManagement      Checking to see if the report files need to be split    {"costmanagement": "Split"}
2020-10-23T15:36:30.388-0400    INFO    controllers.CostManagement      Report files do not require split, generating tar.gz    {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt.tar.gz"}
2020-10-23T15:36:30.389-0400    INFO    controllers.CostManagement      Generated manifest file {"costmanagement": "RenderManifest", "manifest": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-namespace-labels-lookback-202010.csv
6bf0f32a-f5fa-47c0-ad1b-9034a6d77110_openshift_usage_report.0.csv
2020-10-23T15:36:30.389-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-namespace-labels-lookback-202010.csv"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-node-labels-lookback-202010.csv
6bf0f32a-f5fa-47c0-ad1b-9034a6d77110_openshift_usage_report.1.csv
2020-10-23T15:36:30.391-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-node-labels-lookback-202010.csv"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-persistentvolumeclaim-lookback-202010.csv
6bf0f32a-f5fa-47c0-ad1b-9034a6d77110_openshift_usage_report.2.csv
2020-10-23T15:36:30.392-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-persistentvolumeclaim-lookback-202010.csv"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-usage-lookback-202010.csv
6bf0f32a-f5fa-47c0-ad1b-9034a6d77110_openshift_usage_report.3.csv
2020-10-23T15:36:30.393-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-usage-lookback-202010.csv"}
2020-10-23T15:36:30.410-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:36:30.412-0400    INFO    controllers.CostManagement      Created the following files for upload:         {"costmanagement": "Split", "files": ["/tmp/cost-mgmt-operator-reports/data/../cost-mgmt.tar.gz"]}
2020-10-23T15:36:30.412-0400    INFO    controllers.CostManagement      Pausing for 4 seconds before uploading. {"costmanagement": "openshift-cost/costmanagement-sample"}
2020-10-23T15:36:34.412-0400    INFO    controllers.CostManagement      Uploading the following file:   {"costmanagement": "openshift-cost/costmanagement-sample"}
cost-mgmt.tar.gz
2020-10-23T15:36:34.413-0400    INFO    controllers.CostManagement      Request using token authentication      {"costmanagement": "SetupRequest"}
2020-10-23T15:36:34.413-0400    INFO    controllers.CostManagement      Request Headers:        {"costmanagement": "SetupRequest"}
Content-Type [multipart/form-data; boundary=807483f848f494e1669ac5e26bfef64b696192c20a1b04378646c73b7730]
Authorization [Bearer b3BlbnNoaWZ0LXJlbGVhc2UtZGV2K2FiYWlrZW4xZDhnNWVzMHc0Mnhmcnh1bmt5bmpuN253ZWI6ODhYMkc5VU1aWlk0SjBRRktDTzhXT0ZWVldBT1E3UEdYQ1oyVDJIUVpFMkpBWEY5QjJSVjE1UUtIS0dBSUVGWQ==]
User-Agent [cost-mgmt-operator/b172a40e76eddafab73c0231f2e3fb1e94f88cb0 cluster/505a2749-0f02-4ad1-a46a-6216ca3684fc]
HTTP Response Status: 202 Accepted
```
2. Scenario with reports that are larger than the max size (I used nise to generate large reports and then I copied over one to the tmp directory): 
```
ashleybrookeaiken@aaiken-mac korekuta-operator-go % ls -lh /tmp/cost-mgmt-operator-reports/data 
total 816440
-rw-r--r--  1 ashleybrookeaiken  wheel   398M Oct 23 15:39 October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-1.csv
-rw-r--r--  1 ashleybrookeaiken  wheel    70K Oct 20 14:04 cm-openshift-namespace-labels-lookback-202010.csv
-rw-r--r--  1 ashleybrookeaiken  wheel    19K Oct 20 14:04 cm-openshift-node-labels-lookback-202010.csv
-rw-r--r--  1 ashleybrookeaiken  wheel   2.2K Oct 20 14:04 cm-openshift-persistentvolumeclaim-lookback-202010.csv
-rw-r--r--  1 ashleybrookeaiken  wheel   420K Oct 20 14:04 cm-openshift-usage-lookback-202010.csv
ashleybrookeaiken@aaiken-mac korekuta-operator-go % 
```
LOGS: 
```
2020-10-23T15:41:54.502-0400    INFO    controllers.CostManagement      There have been no prior successful uploads to cloud.redhat.com.        {"costmanagement": "checkCycle"}
2020-10-23T15:41:54.502-0400    INFO    controllers.CostManagement      Checking to see if the report files need to be split    {"costmanagement": "Split"}
2020-10-23T15:41:54.503-0400    INFO    controllers.CostManagement      Report files exceed the max size. Splitting files       {"costmanagement": "Split"}
2020-10-23T15:41:54.503-0400    INFO    controllers.CostManagement      Creating file   {"costmanagement": "WritePart", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-11.csv"}
2020-10-23T15:41:55.507-0400    INFO    controllers.CostManagement      Creating file   {"costmanagement": "WritePart", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-12.csv"}
2020-10-23T15:41:57.608-0400    INFO    controllers.CostManagement      Creating file   {"costmanagement": "WritePart", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-13.csv"}
2020-10-23T15:41:58.805-0400    INFO    controllers.CostManagement      Creating file   {"costmanagement": "WritePart", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-14.csv"}
2020-10-23T15:41:59.554-0400    INFO    controllers.CostManagement      Creating file   {"costmanagement": "WritePart", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-15.csv"}
[/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-11.csv /tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-12.csv /tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-13.csv /tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-14.csv /tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-15.csv]
2020-10-23T15:41:59.612-0400    INFO    controllers.CostManagement      Generated manifest file {"costmanagement": "RenderManifest", "manifest": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:41:59.612-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt0.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-11.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:41:59.612-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-11.csv"}
2020-10-23T15:42:01.480-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:01.481-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt1.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-12.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:01.481-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-12.csv"}
2020-10-23T15:42:03.357-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:03.358-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt2.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-13.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:03.359-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-13.csv"}
2020-10-23T15:42:05.169-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:05.169-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt3.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-14.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:05.170-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-14.csv"}
2020-10-23T15:42:07.036-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:07.037-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt4.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-15.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:07.037-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/October-2020-505a2749-0f02-4ad1-a46a-6216ca3684fc-ocp_pod_usage-15.csv"}
2020-10-23T15:42:07.178-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:07.178-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt5.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-namespace-labels-lookback-202010.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:07.178-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-namespace-labels-lookback-202010.csv"}
2020-10-23T15:42:07.180-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:07.181-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt6.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-node-labels-lookback-202010.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:07.181-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-node-labels-lookback-202010.csv"}
2020-10-23T15:42:07.181-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:07.181-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt7.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-persistentvolumeclaim-lookback-202010.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:07.182-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-persistentvolumeclaim-lookback-202010.csv"}
2020-10-23T15:42:07.182-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:07.182-0400    INFO    controllers.CostManagement      Generating tar.gz       {"costmanagement": "Split", "tarFile": "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt8.tar.gz"}
/tmp/cost-mgmt-operator-reports/data/cm-openshift-usage-lookback-202010.csv
33b2f394-375a-48a3-969c-24b8656d1a5a_openshift_usage_report.0.csv
2020-10-23T15:42:07.182-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/cm-openshift-usage-lookback-202010.csv"}
2020-10-23T15:42:07.188-0400    INFO    controllers.CostManagement      Adding file to tar.gz   {"costmanagement": "addFileToTarWriter", "file": "/tmp/cost-mgmt-operator-reports/data/manifest.json"}
2020-10-23T15:42:07.189-0400    INFO    controllers.CostManagement      Created the following files for upload:         {"costmanagement": "Split", "files": ["/tmp/cost-mgmt-operator-reports/data/../cost-mgmt0.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt1.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt2.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt3.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt4.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt5.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt6.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt7.tar.gz", "/tmp/cost-mgmt-operator-reports/data/../cost-mgmt8.tar.gz"]}
2020-10-23T15:42:07.189-0400    INFO    controllers.CostManagement      Pausing for 15 seconds before uploading.        {"costmanagement": "openshift-cost/costmanagement-sample"}
```

## TODO: 
* Run our old python script on the same directory and make sure that the tar.gzs and split csv files are the same 
* Ensure that the CSV header is in each of the split files that I am generating (I checked and it is not so I am fixing that currently) 
* Figure out the flow for where we want to put the tar.gzs/ (delete that directory before splitting files to clean it out each time /or after) and figure out if we should delete the files in the data directory after uploading them. 
* add any errors that happened during the attempted split to the CR status possibly & improve error handling

### CODE REVIEW

File: packaging/packaging.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	_, err = io.Copy(tarWriter, file)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func WriteTarball(logger logr.Logger, tarFileName, manifestFileName, manifestUUID string, archiveFiles []string) string {
+	if len(archiveFiles) > 0 {
```
Comment (maskarb) [inline packaging/packaging.go:156]:
```suggestion
	if len(archiveFiles) <= 0 {
		return ""
	}
```
and then un-nest what is below here

File: packaging/packaging.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+	log.Info("Creating file ", "file", splitFileName)
+	splitFile, err := os.Create(splitFileName)
+	if err != nil {
+		fmt.Println("An error occurred:", err)
+	}
+	// Create the csv writer
+	writer := csv.NewWriter(splitFile)
+	// err = writer.WriteRow(csvHeader)
+	for {
+		row, err := csvReader.Read()
```
Comment (maskarb) [inline packaging/packaging.go:205]: Are there other errors here we should handle?

File: packaging/packaging.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+		if sizeEstimate >= int(maxBytes) {
+			return splitFileName, false
+		}
+	}
+}
+
+func SplitFiles(logger logr.Logger, filePath string) {
+	fileList, err := ioutil.ReadDir(filePath)
+	if err != nil {
+		fmt.Println("could not read dir")
```
Comment (maskarb) [inline packaging/packaging.go:222]: What happens when we can't read the directory?

File: packaging/packaging.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+func Split(logger logr.Logger, filePath string, cost *costmgmtv1alpha1.CostManagement) {
+	log := logger.WithValues("costmanagement", "Split")
+	var outFiles []string
+	log.Info("Checking to see if the report files need to be split")
+	needSplit := NeedSplit(filePath)
+	if needSplit {
+		log.Info("Report files exceed the max size. Splitting files")
+		SplitFiles(logger, filePath)
+		tarPath := filePath + "/../"
```
Comment (maskarb) [inline packaging/packaging.go:264]: Can we use absolute paths instead?

File: packaging/packaging.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+				tarFileName := tarPath + tarFileTmpl + strconv.Itoa(idx) + ".tar.gz"
+				log.Info("Generating tar.gz", "tarFile", tarFileName)
+				outputTar := WriteTarball(logger, tarFileName, manifestFileName, manifestUUID, singleFile)
+				if outputTar != "" {
+					outFiles = append(outFiles, outputTar)
+				}
+			}
+		}
+	} else {
+		tarFileName := filePath + "/../cost-mgmt.tar.gz"
```
Comment (maskarb) [inline packaging/packaging.go:281]: Same as above, can we use absolute paths?

**Review by maskarb — COMMENTED** [review]
Initial comments.

We should probably propagate errors up to the reconciler.

File: packaging/packaging.go (hunk 2 above)
Comment (abaiken) [reply packaging/packaging.go:227]: Added error handling below. Now we catch the EOF error and treat it as the end of file, but if a different error occurs, we bubble it up to the reconciler.

File: packaging/packaging.go (hunk 3 above)
Comment (abaiken) [reply packaging/packaging.go]: This error is now bubbled up to the reconciler and we do not proceed if we can't read the directory.

File: packaging/packaging.go (hunk 4 above)
Comment (abaiken) [reply packaging/packaging.go]: done

File: packaging/packaging.go (hunk 5 above)
Comment (abaiken) [reply packaging/packaging.go]: done

**Review by maskarb — APPROVED** [review]
lgtm!

**Pull request merged**